### ISSUE TITLE
Race condition in JavacPlugin

### ISSUE THREAD

**Issue by iadcode** [manifold-systems/manifold#173]
[`JavacPlugin.loadJavacParserClass()`](https://github.com/manifold-systems/manifold/blob/v2020.1.4/manifold-core-parent/manifold/src/main/java/manifold/internal/javac/JavacPlugin.java#L1098) checks if the classloader contains `ManJavacParser`, and if not, adds it.

When building with [SBT](https://www.scala-sbt.org/), this method is invoked multiple times, sometimes concurrently, which can result in an "attempted duplicate class definition" error:
```
java.lang.LinkageError: loader 'app' (instance of jdk.internal.loader.ClassLoaders$AppClassLoader) attempted duplicate class definition for com.sun.tools.javac.parser.ManJavacParser.
    at java.base/java.lang.ClassLoader.defineClass1(Native Method)
    ...
```

It looks like `loadJavacParserClass()` should be acquiring a lock first. Adding a `synchronized(classLoader) {...}` around the body of this method appears to prevent the error.

**Comment by rsmckinney** [comment]
@iadcode Thanks for reporting this! Fix will be available in next release coming soon.

**Comment by iadcode** [comment]
Great, thanks @rsmckinney !

**Comment by rsmckinney** [comment]
Fixed in release 2020.1.7

**Comment by iadcode** [comment]
Hi @rsmckinney,

Thanks for the fix. This might be hard to believe, but my problem with SBT persists...

I appear to have two threads inside the new `synchronized(JavacPlugin.class){}` block at the same time, yet they're still contending for the same classloader. I don't know exactly what kind of classloader voodoo goes on in the SBT build process, but I imagine it's responsible for this.

I don't know a huge amount about `ClassLoader`, but I see it has a protected `getClassLoadingLock()` method which looks like it's tailored for this use case (and given that we're already hacking away at the protected API, it seems appropriate). Do you foresee any problems with locking on the result of that instead?

I can reliably fix my error by replacing:

    synchronized(JavacPlugin.class)
    {

with:

    synchronized(
      ReflectUtil.method(JavacParser.class.getClassLoader(), "getClassLoadingLock", String.class)
        .invoke("com.sun.tools.javac.parser.ManJavacParser"))
    {

Would you like me to raise another issue for this?

Thanks!

**Comment by rsmckinney** [comment]
@iadcode I don't have any experience with SBT, however using the class lock is probably okay for this use-case. FWIW, SBT is the first reported case of this kind for Manifold. It could be that this issue exists in Maven and/or Gradle too, but that no one uses Manifold in parallel builds. I'll experiment.

Btw, assuming your project uses Scala, the next Manifold release will support Scala and Kotlin as well as Android. It's almost there.

**Comment by shoaniki** [comment]
FWIW I have just seen the exact error message in Gradle when performing a parallel build with the latest Manifold 2021.1.12.  It's only happened once, but I guess that's what you'd expect for a race condition?

**Comment by rsmckinney** [comment]
@shoaniki thanks for reporting this, manifold needs more testing w parallel build
